staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go

```diff
@@ -106,31 +111,48 @@ func parseEndpoint(endpoint string) (string, error) {
 	return u.Path, nil
 }
 
-// Check the KMS provider API version.
-// Only matching kmsapiVersion is supported now.
-func checkAPIVersion(kmsClient kmsapi.KeyManagementServiceClient) error {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
-	defer cancel()
+func (g *gRPCService) checkAPIVersion(ctx context.Context) error {
+	g.mux.Lock()
+	defer g.mux.Unlock()
 
 	request := &kmsapi.VersionRequest{Version: kmsapiVersion}
-	response, err := kmsClient.Version(ctx, request)
+	response, err := g.kmsClient.Version(ctx, request)
 	if err != nil {
 		return fmt.Errorf("failed get version from remote KMS provider: %v", err)
 	}
 	if response.Version != kmsapiVersion {
-		return fmt.Errorf("KMS provider api version %s is not supported, only %s is supported now",
-			response.Version, kmsapiVersion)
+		return fmt.Errorf(versionErrorf, response.Version, kmsapiVersion)
 	}
+	g.versionChecked = true
 
-	glog.V(4).Infof("KMS provider %s initialized, version: %s", response.RuntimeName, response.RuntimeVersion)
+	glog.V(4).Infof("Version of KMS provider is %s", response.Version)
 	return nil
 }
 
+func (g *gRPCService) checkAPIVersionDecorator(ctx context.Context, payload []byte,
+	f func(context.Context, []byte) ([]byte, error)) ([]byte, error) {
+
+	g.mux.RLock()
+	versionChecked := g.versionChecked
+	g.mux.RUnlock()
+
+	if !versionChecked {
+		if err := g.checkAPIVersion(ctx); err != nil {
```

Comment:
I still don't really get what the point of RLocking is, it seems to mean that checkAPIVersion can be called multiple times. That doesn't appear to be harmful (the individual calls are serialized), just wasteful. Maybe you can fix in a follow up if you agree.

Reply:
Will fix this in a follow-up.